### DISCUSSION TITLE
MET installation error - ERROR: Command returned with non-zero (2) status: make  > met.make.log 2>&1

### DISCUSSION THREAD

**Discussion by anwe5** [dtcenter/METplus#2629] [Installation] [answered]
Hi all,

I want to install MET on Levante (dkrz). I encounter some problems during the installation process. I followed the instructions from the user guide to install MET and created an environment configuration file (see `install_met_env.levante` attached). I found a similiar discussion here that suggested to add the compilers version, but that didn't resolve my problem.

These are the last two lines of the output when compiling MET:
```
make > met.make.log 2>&1
ERROR: Command returned with non-zero (2) status: make  > met.make.log 2>&1
```
and this is the output of the `met.make.log` file:
```
collect2: error: ld returned 1 exit status
make[5]: *** [Makefile:434: ensemble_stat] Error 1
make[5]: Leaving directory '/home/b/b382378/MET/MET_releases/11.1.0/MET-11.1.1/src/tools/core/ensemble_stat'
make[4]: *** [Makefile:379: all-recursive] Error 1
make[4]: Leaving directory '/home/b/b382378/MET/MET_releases/11.1.0/MET-11.1.1/src/tools/core'
make[3]: *** [Makefile:373: all-recursive] Error 1
make[3]: Leaving directory '/home/b/b382378/MET/MET_releases/11.1.0/MET-11.1.1/src/tools'
make[2]: *** [Makefile:370: all-recursive] Error 1
make[2]: Leaving directory '/home/b/b382378/MET/MET_releases/11.1.0/MET-11.1.1/src'
make[1]: *** [Makefile:428: all-recursive] Error 1
make[1]: Leaving directory '/home/b/b382378/MET/MET_releases/11.1.0/MET-11.1.1'
make: *** [Makefile:369: all] Error 2
```

Here you can find the log-files, the installation file and the entire screen output from the run of the script: 
[config.log](https://github.com/user-attachments/files/16082342/config.log)
[met.make.log](https://github.com/user-attachments/files/16082345/met.make.log)
[output.txt](https://github.com/user-attachments/files/16082346/output.txt)
[install_met_env_levante.txt](https://github.com/user-attachments/files/16082352/install_met_env_levante.txt)

Thank you so much in advance! Please let me know if you need further files or information.

**Comment by jprestop** [comment]
@anwe5 Thank you for taking the time to look at the User's Guide and at some of the past discussions.  Thank you also for passing along all of these helpful files.

In your install_met_env_levante.txt file, I see:
```
export MET_PYTHON_LD=-L/sw/spack-levante/mambaforge-22.9.0-2-Linux-x86_64-kptncg/lib/python3.10/config-3.10-x86_64-linux-gnu
```
Unfortunately, that line does not contain enough information and it's missing other important information like, for example:
```
 -lpython3.10 -lpthread -ldl  -lutil -lm
```
So, first, let's have you try to find the `python3-config` executable file by running:
```
find /sw/spack-levante/python-3.10.10-p24a46 -name python3-config
```
Hopefully, that command will return a path.  Provide that full path in place of `<full path>` below and run this command:
```
<full path>/python3-config  --ldflags --embed
```
and send the output of that command to us please. We'll take a look and give you a new value to use for MET_PYTHON_LD.

**Reply by anwe5** [reply]
Here's the output:
```
[b382378@levante5 ~]$ find /sw/spack-levante/python-3.10.10-p24a46 -name python3-config
/sw/spack-levante/python-3.10.10-p24a46/bin/python3-config
[b382378@levante5 ~]$ /sw/spack-levante/python-3.10.10-p24a46/bin/python3-config  --ldflags --embed
 -L/sw/spack-levante/python-3.10.10-p24a46/lib -lpython3.10 -lcrypt -lintl -lpthread -ldl  -lutil -lm -lm
```

**Reply by jprestop** [reply]
That output looks great.  Please go ahead and set the following in your install_met_env_levante.txt file:
```
export MET_PYTHON_LD="-L/sw/spack-levante/python-3.10.10-p24a46/lib -lpython3.10 -lcrypt -lintl -lpthread -ldl  -lutil -lm -lm"
```
Since you've already compiled the external libraries, please also uncomment the following lines by removing the "#" in front so that you don't have to wait for them to recompile:
```
#export MET_GRIB2CLIB=/home/b/b382378/wgrib2-3.1.1-4yhjk5/lib
#export MET_GRIB2CINC=/home/b/b382378/wgrib2-3.1.1-4yhjk5/include
#export GRIB2CLIB_NAME=libwgrib2.a
#export MET_BUFRLIB=/sw/spack-levante/bufr-11.7.1-bqujfc/lib64
#export BUFRLIB_NAME=libbufr_4.a
```
And, I see that you have the following set:
```
export MET_PYTHON_BIN_EXE=/sw/spack-levante/mambaforge-22.9.0-2-Linux-x86_64-kptncg/lib/python3.10/
```
However, I would be surprised if the Python executable was in that location. I think the value you want to set it to is likely `/sw/spack-levante/python-3.10.10-p24a46/bin/python3`:
```
export MET_PYTHON_BIN_EXE=/sw/spack-levante/python-3.10.10-p24a46/bin/python3
```
Please let us know if you have any questions.  After you make these changes, please rerun and let us know how it goes.

**Reply by anwe5** [reply]
Thank you for taking the time to look into my issue. I followed your instructions and reran the script. Unfortunately, I still get an error message: `ERROR: Command returned with non-zero (2) status: make  > met.make.log 2>&1`

Here is the output from the run of the script and the install_met_env_levante-file. I also uncommend `export MET_PYTHON=/sw/spack-levante/python-3.10.10-p24a46` since I figured that the script was otherwise loading the wrong Python version (3.6.8 instead of 3.10.10).

[config.log](https://github.com/user-attachments/files/16140547/config.log)
[install_met_env_levante.txt](https://github.com/user-attachments/files/16140548/install_met_env_levante.txt)
[met.make.log](https://github.com/user-attachments/files/16140549/met.make.log)
[output.txt](https://github.com/user-attachments/files/16140550/output.txt)

**Reply by jprestop** [reply]
@anwe5 Thank you for providing these files to us.  I see the following error in your met.make.log file:
```
In file included from /sw/spack-levante/mambaforge-22.9.0-2-Linux-x86_64-kptncg/include/python3.10/Python.h:36,
                 from ../../../src/libcode/vx_python3_utils/python3_numpy.h:15,
                 from ../../../src/libcode/vx_python3_utils/vx_python3_utils.h:13,
                 from ../../../src/basic/vx_util/observation.h:28,
                 from python_pointdata.cc:14:
/usr/include/unistd.h:779:12: error: conflicting declaration of ‘int isatty(int)’ with ‘C’ linkage
 extern int isatty (int __fd) __THROW;
            ^~~~~~
In file included from ../../../src/basic/vx_util/observation.h:25,
                 from python_pointdata.cc:14:
../../../src/basic/vx_util/util_constants.h:138:15: note: previous declaration with ‘C++’ linkage
    inline int isatty(int fd) { return 0; }
               ^~~~~~
```
and the following error at the *isatty* check:
```
configure:9897: checking for isatty
configure:9897: /usr/bin/gcc -o conftest -g -O2  -fopenmp -I/home/b/b382378/MET/netcdf/include -I${MET_HDF5}/include -I/home/b/b382378/wgrib2-3.1.1-4yhjk5/include -DMET_PYTHON_BIN_EXE="\"${MET_PYTHON_BIN_EXE}\"" -I../../basic/vx_log -I../../basic/vx_util -I/sw/spack-levante/mamb\
aforge-22.9.0-2-Linux-x86_64-kptncg/include/python3.10 -I${MET_GSL}/include -DBLOCK4 -DWITH_GRIB2 -DWITH_PYTHON -DMET_BASE="\"$(pkgdatadir)\"" -Wl,--disable-new-dtags -Wl,-rpath,/home/b/b382378/MET/MET_releases/11.1.0/external_libs/lib:/home/b/b382378/MET/MET_releases/11.1.0/ext\
ernal_libs/lib:/home/b/b382378/MET/MET_releases/11.1.0/external_libs/lib:/sw/spack-levante/bufr-11.7.1-bqujfc/lib64:/home/b/b382378/wgrib2-3.1.1-4yhjk5/lib:/sw/spack-levante/python-3.10.10-p24a46/lib:/sw/spack-levante/gsl-2.7-5ou4yk/lib -Wl,-rpath,/usr/lib64/sw/spack-levante/zli\
b-1.2.13-7izaze/lib -L/usr/lib64  -L/home/b/b382378/MET/MET_releases/11.1.0/external_libs/lib  -fopenmp -L/home/b/b382378/MET/netcdf/lib -L${MET_HDF5}/lib -L/home/b/b382378/wgrib2-3.1.1-4yhjk5/lib -L/sw/spack-levante/python-3.10.10-p24a46/lib -lpython3.10 -lcrypt -lintl -lpthrea\
d -ldl  -lutil -lm -lm -L${MET_GSL}/lib -L/sw/spack-levante/bufr-11.7.1-bqujfc/lib64 conftest.c  -lhdf5_hl -lhdf5 -lz >&5
/usr/bin/ld: cannot find -lintl
collect2: error: ld returned 1 exit status
```
Since there is an error with `-lintl` not being found, we'll try removing it for now.  It may be the case that `-lintl` is needed but it may not be needed.  Let's have you try modifying the value of MET_PYTHON_LD in your install_met_env_levante.txt file to be:
```
export MET_PYTHON_LD="-L/sw/spack-levante/python-3.10.10-p24a46/lib -lpython3.10 -lcrypt -lpthread -ldl  -lutil -lm -lm"
```
After you make this change, please rerun and let us know how it goes.

**Reply by anwe5** [reply]
I changed the MET_PYTHON_LD variable, as you suggested. Now there is new error in the met.make.log connected to MET_GRIB2CLIB: 
```
g++: error: libwgrib2.a: No such file or directory
make[5]: *** [Makefile:434: ensemble_stat] Error 1
make[5]: Leaving directory '/home/b/b382378/MET/MET_releases/11.1.0/MET-11.1.1/src/tools/core/ensemble_stat'
make[4]: *** [Makefile:379: all-recursive] Error 1
make[4]: Leaving directory '/home/b/b382378/MET/MET_releases/11.1.0/MET-11.1.1/src/tools/core'
make[3]: *** [Makefile:373: all-recursive] Error 1
make[3]: Leaving directory '/home/b/b382378/MET/MET_releases/11.1.0/MET-11.1.1/src/tools'
make[2]: *** [Makefile:370: all-recursive] Error 1
make[2]: Leaving directory '/home/b/b382378/MET/MET_releases/11.1.0/MET-11.1.1/src'
make[1]: *** [Makefile:428: all-recursive] Error 1
make[1]: Leaving directory '/home/b/b382378/MET/MET_releases/11.1.0/MET-11.1.1'
make: *** [Makefile:369: all] Error 2
```
I double-checked the path for MET_GRIB2CLIB and the filename, and both are correct. However, I had to install GRIB2, so I am not sure if I might have installed the wrong version.

Again, thanks for taking the time to look into my issue :)

**Reply by anwe5** [reply]
Forgott to attach these:
[config.log](https://github.com/user-attachments/files/16159212/config.log)
[install_met_env_levante.txt](https://github.com/user-attachments/files/16159214/install_met_env_levante.txt)
[met.make.log](https://github.com/user-attachments/files/16159217/met.make.log)
[output.txt](https://github.com/user-attachments/files/16159219/output.txt)

**Reply by jprestop** [reply, marked as answer]
Hi @anwe5.  Thank you again for the attached files.  Ah, yes, wgrib2 is not the library that is needed.  Rather, it is the g2clib library that is needed.

The compile_MET_all.sh script will install this library and its dependencies (libjasper.a, libpng.a, and libz.a) for you, if you'd like.  What you would need to do is download the tar files that we provide and place them in the tar_files directory:
```
wget https://dtcenter.ucar.edu/dfiles/code/METplus/MET/installation/tar_files.tgz
```
Note that the MET tar file (v11.1.1.tar.gz) is NOT included here, but you already have that.

In your install_met_env_levante.txt file, you would comment out the following lines by adding a `#` in front:
```
export MET_GRIB2CLIB=/home/b/b382378/wgrib2-3.1.1-4yhjk5/lib
export MET_GRIB2CINC=/home/b/b382378/wgrib2-3.1.1-4yhjk5/include
export GRIB2CLIB_NAME=libwgrib2.a
export LIB_JASPER=/usr/lib64
export LIB_PNG=/usr/lib64
export LIB_Z=/sw/spack-levante/zlib-1.2.13-7izaze/lib
```
The libraries and their header files would be installed lib and include subdirectories, respectively, of the following directory:
```
/home/b/b382378/MET/MET_releases/11.1.0/external_libs
```
I see that you currently have BUFRLIB_NAME set to `libbufr_4.a`.  But, instead, you want that to be set to `-lbufr_4` if the name of the library file is `libbufr_4.a`.  Also, should you want to reinstall MET after the grib2clib library is installed, and not want to reinstall grib2clib and its dependencies, you would then set the following values in your install_met_env_levante.txt file:
```
export MET_GRIB2CLIB=/home/b/b382378/MET/MET_releases/11.1.0/external_libs/lib
export MET_GRIB2CINC=/home/b/b382378/MET/MET_releases/11.1.0/external_libs/include
export GRIB2CLIB_NAME=-lgrib2c
export LIB_JASPER=/home/b/b382378/MET/MET_releases/11.1.0/external_libs/lib
export LIB_PNG=/home/b/b382378/MET/MET_releases/11.1.0/external_libs/lib
export LIB_Z=/home/b/b382378/MET/MET_releases/11.1.0/external_libs/lib
```
I hope that helps.  After you make this change, please rerun and let us know how it goes.

**Reply by anwe5** [reply]
This resolved the error, and MET has been installed successfully. Thank you for your help!

**Reply by jprestop** [reply]
That's great news @anwe5!  Thank you for letting us know for you ended up with a successful compilation.
